### PR TITLE
`forum-search`: fix addon by using new ScratchDB API

### PR DESCRIPTION
**Pull request by GrahamSH-LLK**
i submit quality PRs
fixes the forum search (switches to meilisearch)
uses a separate input for the filtering now because that's how meilisearch likes it

### CODE REVIEW

**Comment by WorldLanguages** [comment]
@GrahamSH-LLK We need to remove the notice from addon.json which says `The ScratchDB database used by this addon is offline, so this addon does not work.`

File: addons/forum-search/userscript.js [hunk 1]
```diff
@@ -255,21 +266,24 @@ export default async function ({ addon, console, msg }) {
...
         break;
       }
     }
   }
 
   searchBar.setAttribute("placeholder", searchPlaceholder);
+  filterBar.setAttribute("placeholder", msg("filter-placeholder"));
   search.appendChild(searchBar);
+  search.appendChild(filterBar);
+  search.appendChild(Object.assign(document.createElement("input"), { type: "submit", style: "display: none" })); // this is a hack to make the enter key work
```
Comment (mxmou) [inline addons/forum-search/userscript.js:286]: I actually think the button should be visible and labelled "Search" - it's weird that there currently isn't one.

File: addons/forum-search/userscript.js [hunk 2]
```diff
@@ -150,15 +161,17 @@ function appendSearch(box, query, page, term, msg) {
...
         postLeftDl.appendChild(document.createElement("br"));
-
-        if (locationQuery !== "") {
+        const filterBar = document.querySelector("#forum-search-filter-input");
+        if (filterBar.value !== "") {
           let userPostButton = document.createElement("a");
           userPostButton.appendChild(document.createTextNode(msg("posts-here")));
           userPostButton.addEventListener("click", () => {
-            document.getElementById("forum-search-input").value = `+username:"${post.username}" ${locationQuery}`;
+            filterBar.value = `username = "${post.username}" AND ${filterBar.value}`;
```
Comment (mxmou) [inline addons/forum-search/userscript.js:169]:
```suggestion
            document.getElementById("forum-search-input").value = "";
            filterBar.value = `username = "${post.username}" AND ${filterBar.value}`;
```

File: addons/forum-search/userscript.js [hunk 3]
```diff
@@ -171,10 +184,11 @@ function appendSearch(box, query, page, term, msg) {
         let userGlobalButton = document.createElement("a");
         userGlobalButton.appendChild(document.createTextNode(msg("posts-sitewide")));
         userGlobalButton.addEventListener("click", () => {
-          document.getElementById("forum-search-input").value = `+username:"${post.username}"`;
+          document.getElementById("forum-search-filter-input").value = `username = "${post.username}"`;
```
Comment (mxmou) [inline addons/forum-search/userscript.js:187]:
```suggestion
          document.getElementById("forum-search-input").value = "";
          document.getElementById("forum-search-filter-input").value = `username = "${post.username}"`;
```

**Review by mxmou — COMMENTED** [review]
The messages `first-checked` and `last-checked` and the `locationQuery` variable are now unused.

The different placeholders ("Search posts in <category/topic>") don't really make sense after these changes because the user can change or remove the default filter. I think the input should just say "Search posts".

Is the beta tag still needed? It was added when ScratchDB broke.

File: addons/forum-search/userscript.js (hunk 1 above)
Comment (mxmou) [reply addons/forum-search/userscript.js:280]: You can now remove the invisible button.

**Review by mxmou — APPROVED** [review]

File: addons-l10n/en/forum-search.json [hunk 4]
```diff
@@ -5,13 +5,10 @@
...
   "forum-search/last-edited-by": "Last edited by {username} ({datetime})",
-  "forum-search/placeholder": "Search posts on the entire Scratch Forums",
-  "forum-search/search-topic": "Search posts in \"{topic}\"",
-  "forum-search/search-cat": "Search posts in \"{cat}\"",
+  "forum-search/placeholder": "Search posts",
   "forum-search/relevance": "relevance",
   "forum-search/newest": "newest",
-  "forum-search/oldest": "oldest"
+  "forum-search/oldest": "oldest",
+  "forum-search/filter-placeholder": "Filter (eg. username = \"\")"
```
Comment (WorldLanguages) [inline addons-l10n/en/forum-search.json:13]:
```suggestion
  "forum-search/filter-placeholder": {
    "string": "Filter (e.g. username = \"\")",
    "comment": "Do not translate 'username' word"
  }
```

(`eg.` -> `e.g.`)

**Comment by WorldLanguages** [comment]
> You can now remove the invisible button.

@mxmou What button are you talking about? You may push to this branch, if you want.

File: addons-l10n/en/forum-search.json [hunk 5]
```diff
@@ -5,13 +5,13 @@
...
-  "forum-search/search-topic": "Search posts in \"{topic}\"",
-  "forum-search/search-cat": "Search posts in \"{cat}\"",
+  "forum-search/placeholder": "Search posts",
   "forum-search/relevance": "relevance",
   "forum-search/newest": "newest",
-  "forum-search/oldest": "oldest"
+  "forum-search/oldest": "oldest",
+  "forum-search/filter-placeholder": {
+    "string": "Filter (e.g. username = \"\")",
+    "comment": "Do not translate 'username' word"
```
Comment (WorldLanguages) [inline addons-l10n/en/forum-search.json:15]: Whoops

```suggestion
    "developer_comment": "Do not translate 'username' word"
```

**Comment by GrahamSH-LLK** [comment]
@mxmou the new button styling looks weird with r2 enabled, i think `height: revert;` should work instead of the explicit height?

File: addons/forum-search/userscript.js [hunk 6]
```diff
@@ -56,30 +55,41 @@ function cleanPost(post) {
...
-    .fetch(`https://scratchdb.lefty.one/v3/forum/search?q=${encodeURIComponent(query)}&page=${page}&o=${term}`)
+    .fetch(
+      `https://scratchdb.lefty.one/search/indexes/forum_posts/search?attributesToSearchOn=content&hitsPerPage=50&q=${encodeURIComponent(
+        query
+      )}&filter=${encodeURIComponent(filter)}&page=${page + 1}${
+        term == "newest" ? "&sort=id:desc" : term == "oldest" ? "&sort=id:asc" : ""
+      }`,
+      {
+        headers: {
+          authorization: "Bearer 3396f61ef5b02abf801096be5f0b0ee620de304dd92fc6045aeb99539cd0bec4",
```
Comment (WorldLanguages) [inline addons/forum-search/userscript.js:82]: Adding note about this token being public

```suggestion
          // Note: the following token is well-known and public.
          authorization: "Bearer 3396f61ef5b02abf801096be5f0b0ee620de304dd92fc6045aeb99539cd0bec4",
```

**Review by WorldLanguages — APPROVED** [review]

**Pull request merged**